server/gokbg3/grails-app/services/org/gokb/ValidationService.groovy

```diff
@@ -0,0 +1,610 @@
+      mandatory: false
+    ],
+    publication_type: [
+      mandatory: true
+    ],
+    date_monograph_published_print: [
+      mandatory: false,
+      pubType: "Monograph",
+      validator: [
+        name: "checkDate",
+        args: []
+      ]
+    ],
+    date_monograph_published_online: [
+      mandatory: false,
+      pubType: "Monograph",
+      validator: [
+        name: "checkDate",
+        args: []
+      ]
+    ],
+    monograph_volume: [
+      mandatory: false,
+      pubType: "Monograph"
+    ],
+    monograph_edition: [
+      mandatory: false,
+      pubType: "Monograph"
+    ],
+    first_editor: [
+      mandatory: false,
+      pubType: "Monograph"
+    ],
+    parent_publication_title_id: [
+      mandatory: false
+    ],
+    preceding_publication_title_id: [
+      mandatory: false
+    ],
+    access_type: [
+      mandatory: false,
+      validator: [
+        name: "checkAccessType",
+        args: []
+      ]
+    ],
+    zdb_id: [
+      mandatory: false,
+      namespaces: [
+        'Serial': 'zdb',
+      ],
+      validator: [
+        name: "checkKbartIdentifier",
+        args: ["_colName", "publication_type"]
+      ]
+    ]
+  ]
+
+  def generateKbartReport(InputStream kbart, IdentifierNamespace titleIdNamespace = null) {
+    def result = [
+      valid: true,
+      rows: [total: 0, error: 0, warning: 0],
+      errors: [
+        missingColumns: [],
+        rows: [:],
+        type: [:]
+      ],
+      warnings: [
+        missingColumns: [],
+        rows: [:],
+        type: [:]
+      ]
+    ]
+
+    CSVReader csv = initReader(kbart)
+
+    Map col_positions = [:]
+    String[] header = csv.readNext()
+
+    header = header.collect { it.toLowerCase().trim() }
+
+    int ctr = 0
+
+    header.each {
+      col_positions[it] = ctr++
+    }
+
+    knownColumns.each { colName, info ->
+      if (info.mandatory && !header.contains(colName)) {
+        result.errors.missingColumns.add(colName)
+      }
+      else if (!header.contains(colName) && colName != 'zdb_id') {
+        result.warnings.missingColumns.add(colName)
+      }
+    }
+
+    String[] nl = csv.readNext()
+    int rowCount = 0
+
+    if (result.errors.missingColumns.size() == 0) {
+      while (nl != null) {
+        rowCount++
+
+        if (nl.size() > 1 && nl.size() != header.size()) {
+          result.rows.total++
+          result.rows.error++
+          result.errors.rows["${rowCount}"] = [
+            columnsCount: [
+              message: "Inconsistent column count (${nl.size()} <> ${header.size()})!",
+              messageCode: "kbart.errors.tabsCountRow",
+              args: [nl.size(), header.size()]
+            ]
+          ]
+          if (!result.errors.type['columnCount']) {
+            result.errors.type['columnCount'] = 1
+          }
+          else {
+            result.errors.type['columnCount']++
+          }
+          result.valid = false
+        }
+        else if (nl.size() >= 5) {
```

Comment:
Instead of `5` we could add a static variable like `NUM_MAND_COLUMNS` which is statically initialized after the initialization of `knownColumns`.